cloud-checksum/src/lib.rs

```diff
@@ -60,74 +57,157 @@ impl Commands {
+    #[arg(short, long, env, default_value = "equality")]
+    pub group_by: GroupBy,
+}
+
+/// The tag mode to use when copying files.
+#[derive(Debug, Clone, ValueEnum, Copy, Default)]
+pub enum MetadataCopy {
+    #[default]
+    /// Copy all tags or metadata and fail if it could not be copied.
+    Copy,
+    /// Do not copy any tags or metadata.
+    Supress,
+    /// Attempt to copy tags or metadata but do not fail if it could not be copied.
+    BestEffort,
+}
+
+impl MetadataCopy {
+    /// Is this a copy metadata operation.
+    pub fn is_copy(&self) -> bool {
+        matches!(self, MetadataCopy::Copy | MetadataCopy::BestEffort)
+    }
+
+    /// Is this a best-effort copy metadata operation.
+    pub fn is_best_effort(&self) -> bool {
+        matches!(self, MetadataCopy::BestEffort)
+    }
+}
+
+/// Mode to execute copy task in.
+#[derive(Debug, Clone, ValueEnum, Copy, Default)]
+pub enum CopyMode {
+    /// Always use server-side copy operations if they are available. This may still download and
+    /// upload if it is not possible to server-side copy.
+    #[default]
+    ServerSide,
+    /// Download the object first and then upload it to the destination.
+    DownloadUpload,
+}
+
+impl CopyMode {
+    /// Is this a download-upload copy operation.
+    pub fn is_download_upload(&self) -> bool {
+        matches!(self, CopyMode::DownloadUpload)
+    }
+}
+
+/// The copy subcommand components.
+#[derive(Debug, Args)]
+pub struct Copy {
+    /// The source file to copy from. By default, accepts a file name, use - to accept input from
+    /// stdin.
+    #[arg(required = true)]
+    pub source: String,
+    /// The destination to copy files to. If the input contains multiple files, then this must
+    /// be a directory.
+    #[arg(required = true)]
+    pub destination: String,
+    /// Controls how tags and metadata are copied. By default, this will copy all tags and metadata
+    /// and fail if the tags could not be copied.
+    #[arg(long, env, default_value = "copy")]
+    pub metadata_mode: MetadataCopy,
+    #[arg(long, env, default_value = "server-side")]
+    pub copy_mode: CopyMode,
+    /// The threshold at which a file uses multipart uploads when copying to S3. This can be
+    /// specified with a size unit, e.g. 8mib. By default, a multipart copy will occur when the
+    /// source file was uploaded using multipart, in order to match sums. This can be used to
+    /// override that.
+    #[arg(short, long, env, value_parser = |s: &str| parse_size(s))]
+    pub multipart_threshold: Option<u64>,
+    /// The part size to use when copying files using multipart uploads. This can be specified with
+    /// a size unit, e.g. 8mib. By default, the part size will be automatically determined based on
+    /// how the source was uploaded. This can be used to override that.
+    #[arg(short, long, env, value_parser = |s: &str| parse_size(s))]
+    pub part_size: Option<u64>,
+}
```

Comment:
Not at all needed for this PR - but I think the eventual mode we will prefer/have a flag for is where threshold and part size are determined by reference to the details we can derive from the source object itself.

Reply:
Yes that's the intent. I was thinking that this flag would only be used to override any of that automatic detection?